### PR TITLE
stdenv: fix custom hardening settings when using `__structuredAttrs = true;`

### PR DESCRIPTION
**Pull request by Ma27**
Alternative to #353131

cc @emilazy @wolfgangwalther 

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Ma27** [comment]
> pkgs/build-support/cc-wrapper/fortran-hook.sh:: ${NIX_HARDENING_ENABLE="@default_hardening_flags_str@"}
> pkgs/build-support/cc-wrapper/setup-hook.sh:: ${NIX_HARDENING_ENABLE="@default_hardening_flags_str@"}
> pkgs/build-support/bintools-wrapper/setup-hook.sh:: ${NIX_HARDENING_ENABLE="@default_hardening_flags_str@"}

We're only setting defaults here: I think it's OK to continue using strings here until we kill support for non-structuredAttrs (if that ever happens), right?

> pkgs/development/tools/misc/ltrace/default.nix:    NIX_HARDENING_ENABLE="" \

turns off hardening for checkPhase. Same as above, as long as we support strings, that's OK I think.

> pkgs/build-support/bintools-wrapper/add-hardening.sh:for flag in ${NIX_HARDENING_ENABLE_@suffixSalt@-}; do

This is the line we're currently fixing.

> pkgs/build-support/bintools-wrapper/add-flags.sh:    NIX_HARDENING_ENABLE

Hmm, this will indeed break (I think).
Will look into it.

---

Regarding the rest: what's the suggested course of action here? These will need to be touched when structuredAttrs=true is the new default or shall we do that earlier (haven't thought about the how though).

File: pkgs/stdenv/generic/make-derivation.nix [hunk 1]
```diff
@@ -413,7 +413,7 @@ else let
       enableParallelChecking = attrs.enableParallelChecking or true;
       enableParallelInstalling = attrs.enableParallelInstalling or true;
     } // optionalAttrs (hardeningDisable != [] || hardeningEnable != [] || stdenv.hostPlatform.isMusl) {
-      NIX_HARDENING_ENABLE = enabledHardeningOptions;
```
[the remark this answers is not in the document]
Comment (Ma27) [reply pkgs/stdenv/generic/make-derivation.nix:416]: Aren't non-env things also exposed as bash variables in the builder, just not into the environment?
I mean, the NIX_HARDENING_ENABLE part was seen by `add-hardening.sh` before, but itw as an actual array, not a string.

**Comment by kirillrdy** [comment]
can confirm that this PR fixes postgresql performance regression

**Comment by Ma27** [comment]
So, IMHO this is ready now: it has a regression test that broke for me on master and passes on this commit and seems like the least disruptive change (except for the rebuild amount of course).

cc @risicle @emilazy @wolfgangwalther 
also cc @vcunat as fyi that there will be a fat rebuild on staging that needs to land in 24.11.

**Review by wolfgangwalther — APPROVED** [review]
Just finished rebuilding the world - LGTM.

I confirmed the test breaks on master, passes on this branch. I also played around with the test, changing stuff to confirm it works as intended.

I double checked manually, that the `NIX_HARDENING_ENABLE` variable is consistently set with both structuredAttrs and without.

---

I am still building the 4 packages that we know of which set hardening flags manually and have structuredAttrs enabled already:
- librandombytes
- scalapack
- postgresql
- texlive

Build failures here shouldn't stop us from doing this fix, though. We'd need to fix packages, too, but we can proceed with this PR already.

**Comment by risicle** [comment]
This reminded me that I'd really love to get the faster `uniq` parts of  #119286 into lib so that I'd feel comfortable using unique filters more extensively in the hardening flags so we end up with a normalized result in `NIX_HARDENING_ENABLE`. This could avoid rebuilds arising from e.g. specifying a flag in `hardeningEnable` that's already enabled by default or adding one that's not supported on the platform to `hardeningDisable`.

**Comment by vcunat** [comment]
> fyi that there will be a fat rebuild on staging that needs to land in 24.11.

Rebuilding everything is fine.  I think we had a stdenv rebuild in `master..staging` already anyway.  Now it's primarily being careful about possible regressions ;-)

**Pull request merged**